### PR TITLE
Implement gadget monotone

### PR DESCRIPTION
**Pull request by han0110**
In memory circuit, we need to check if `gc` are strict-monotone in the same `address`. And `gc` has some not too large maximum value so we can be sure diff of each two adjacent row will not exceed the maximum (I call `RANGE` here). (#7)

In this gadget, we pass a advice column for check, with a function returning any expression to play as a switch of this gadget. In this way, we can use `is_equal_rot`'s expression to check if `gc` are strict-monotone when `address === address_prev`.

Also the gadget has three const-generic parameters `RANGE`, `INCR` and `STRICT` for flexibility.

- `RANGE` - specifying the maximum the diff could be
- `INCR` - should be monotone increasing or monotone decreasing (`q_enable` always specify `Rotation::cur()` and compare to `Rotation::prev()`)
- `STRICT` - should disallow equality of diff or not.

For `gc` we could use `StrictMonoIncrChip` for convenience.

### CODE REVIEW

File: src/gadget/monotone.rs
Comment (therealyingtong) [inline src/gadget/monotone.rs:12]: Include the advice column in the Config.

```suggestion
pub(crate) struct MonotoneConfig {
    range_table: Column<Fixed>,
    value: Column<Advice>,
}
```

File: src/gadget/monotone.rs
Comment (therealyingtong) [inline src/gadget/monotone.rs:22]: I'm not sure that we need to define these here; we can just leave it up to the caller to instantiate `MonotoneChip` with the const generics they need.

File: src/gadget/monotone.rs
Comment (therealyingtong) [inline src/gadget/monotone.rs:31]: It may be simpler here to just pass in `q_enable: Expression<F>`, which is still general enough for a synthetic selector.

File: src/gadget/monotone.rs
Comment (therealyingtong) [inline src/gadget/monotone.rs:14]:
```suggestion
/// MonotoneChip helps to check if an advice column is monotonically increasing
```

File: src/gadget/monotone.rs
Comment (therealyingtong) [inline src/gadget/monotone.rs:127]: I suggest separating the strict and non-strict tests, so we can identify which is expected to fail in the unit tests.

**Review by therealyingtong — CHANGES_REQUESTED** [review]

File: src/gadget/monotone.rs
Comment (han0110) [reply src/gadget/monotone.rs:31]: I tried to pass `Expression<F>` at first actually, but the only way to create a `Expression` and put the column into `Gate.queried_selectors` is to call `VirtualCells`'s function, and the only way we have `VirtualCells` is to borrow `mut ConstraintSystem` (call `create_gate` or `lookup`), but the first parameter of `configure` is also a borrowed `mut ConstraintSystem`, which will be blocked by borrow checker. So finally I yield to pass a function to get the expression without borrowing `mut ConstraintSystem`.

Maybe there is other better way to pass through this, do you have any suggestion I can try?

File: src/gadget/monotone.rs
Comment (han0110) [reply src/gadget/monotone.rs]: Fixed in third commit

File: src/gadget/monotone.rs
Comment (han0110) [reply src/gadget/monotone.rs]: Fixed in second commit

File: src/gadget/monotone.rs
Comment (han0110) [reply src/gadget/monotone.rs]: Fixed in second commit

File: src/gadget/monotone.rs
Comment (therealyingtong) [reply src/gadget/monotone.rs:29]: Let's use this pattern for now!

**Review by therealyingtong — APPROVED** [review]
LGTM

**Pull request merged**
